### ISSUE TITLE
Marginal case of dead resource feeding

### ISSUE THREAD

**Issue by bradduthie** [ConFooBio/gmse#63]
There appears to be a very marginal case in which culled resources could do some feeding in the following time step, thereby slightly affecting crop production on a single landscape cell. Here's the problem line: https://github.com/ConFooBio/gmse/blob/master/src/resource.c#L557

While the most common case checks to see if a resource has been culled before allowing it to feed on the landscape (https://github.com/ConFooBio/gmse/blob/master/src/resource.c#L401), the old `res_landscape_interaction` (https://github.com/ConFooBio/gmse/blob/master/src/resource.c#L309) function never needed this in versions prior to 0.6 because it used to be at the end of the resource function. 

There needs to be an AND in the `if` statement to check that the resource has not been culled. Lines 333 and 334 also look suspect; I don't think that the columns are correct. This situation is very rare, and would have little to no effect on the system dynamics if it arose, but it needs to be fixed before the next release of GMSE.

**Comment by bradduthie** [comment]
The columns look okay actually, so I'm not sure what I was talking about there. Nevertheless, the marginal case of dead resources feeding under the original system of resource-landscape interactions should now be resolved. I've not changed anything else about this old system of feeding (e.g., increasing the consumption column of the resource) because it's really a legacy way of doing things, and the new `resource_feeding` function should be an improvement. 

I'd actually appreciate any thoughts on this @jejoenje @AdrianBach @MattNuttall00 -- as it stands, there are two types of feeding that lead to two different orders of operation. The old way works like this:

1. Time step added
2. Resources consume on their landscape cell
3. Birth of new resources
4. Death of resources
5. Remaining resources move
6. Check if the population is over carrying capacity (I think this was important for something @AdrianBach is doing, but can't remember? It just puts it into the paras vector but doesn't actually do anything about it).

The new way works like this:

1. Time step added
2. Resources consume on their landscape cell, then move, then consume on the new landscape cell, and then move again, etc., until until the number of feeds they get runs out (i.e., if just 1, then resources just eat and then move to a new cell). An advantage is that this all happens in a random order so that no individual gets to eat first just because it's first in the array (hence partly why I want to encourage this way of doing things).
3. Birth of new resources
4. Death of resources
5. Remaining resources move (again)
6. Check if the population is over carrying capacity

My question is, would it be best to just get rid of the second movement step in the new method? This seems unnecessary. Also, is it worth adding the consumption amount to the resource array in the old way, even though that amount doesn't really do anything in the simulation (e.g., could just be used to see how much each individual ate, but nothing else)?

**Comment by jejoenje** [comment]
First of all the new way looks good to me @bradduthie, I like the mix of moving and feeding, it seems to make sense to me from a realism perspective, and I think could be important in particular if there are many land owning stakeholders - i.e. feeding would potentially impact stakeholders more variably compared to the old way?
In terms of the second movement step... I'm not sure if I follow why it's there in the new order of ops? Is it just the "leftover" from the old way, where the consumption step didn't also involve movement? I would agree that I don't __think__ it would be necessary to keep in, but I guess it does mean that "new" resources would never move in the time step they are born in? Not saying that's an issue, just highlighting the assumption.
Second, adding the consumption amount into the resource array. Am I right in thinking that column is never used at the moment? If so, part of me thinks you might as well add it - from an IBM perspective it may be useful to access under certain circumstances? The data is there, may as well store it?

**Comment by bradduthie** [comment]
Thanks @jejoenje -- I agree that the variance in feeding is probably more realistic. I've kept the feed, move, feed, etc., pattern, but added an if statement [here](https://github.com/ConFooBio/gmse/commit/bebffc0f4b89ab1aca685e5b709d87a29ce05fe7) so that resources don't move twice after feeding their final time. The end result, I think, is that individuals move a more realistic number of times, but that offspring also move after birth. A lot of this was to avoid the situation in which individuals moved before feeding, which was problematic because it meant that 'scaring' was sometimes redundant given sufficient dispersal distance (why scare when the resource is just going to move at the start of the next time step before feeding anyway?). I think the end result makes more sense, and the explanation is now in one of the vignettes.

I've added the consumption amount to the resource array under the old system of resource landscape interactions -- agree that we might as well store it, even if it's likely never to be used. Closing this issue now.
